### ISSUE TITLE
child_spec warnings with Elixir 1.6

### ISSUE THREAD

**Issue by micahrye** [jeremyjh/dialyxir#112]
Not sure if this counts as an issue since Elixir 1.6 is still 1.6.0-rc.0. 

### Using Elixir 1.6.0-rc.0
The issue is that when running 1.6.0-rc.0 dialyzer will produce warrnings if you do not explicitly define a `child_spec` when you use a GenServer, `use GenServer`. For example, given the following module that uses GenServer: 
```
defmodule GenExample do
  use GenServer

  def start_link(_args) do
    GenServer.start_link(__MODULE__, [], name: __MODULE__)
  end

  def init(state) do
    {:ok, state}
  end
end
```

If we run `$ mix dialyzer` we get the following: 
```
mix dialyzer
Compiling 1 file (.ex)
Checking PLT...
[:compiler, :elixir, :kernel, :logger, :stdlib]
PLT is up to date!
Starting Dialyzer
dialyzer args: [
  check_plt: false,
  init_plt: '/Users/micahrye/code/work/tulip/projects/elixir/sysfs/_build/dev/dialyxir_erlang-20.0_elixir-1.6.0-rc.0_deps-dev.plt',
  files_rec: ['/Users/micahrye/code/work/tulip/projects/elixir/sysfs/_build/dev/lib/sysfs/ebin'],
  warnings: [:unknown]
]
done in 0m1.12s
lib/gen_server.ex:2: Function child_spec/1 has no local return
lib/gen_server.ex:581: The call 'Elixir.Supervisor':child_spec(_@2::#{'id':='Elixir.GenExample', 'start':={'Elixir.GenExample','start_link',[any(),...]}},[]) breaks the contract (child_spec() | {module(),arg::term()} | module(),elixir:keyword()) -> child_spec()
done (warnings were emitted)
```

If we add a child_spec to our example module above: 
```
defmodule GenExample do
  use GenServer

  def child_spec(arg) do
    %{
      id: __MODULE__,
      start: {__MODULE__, :start_link, [arg]},
      restart: :permanent,
      shutdown: 5000,
      type: :worker
    }
  end
  
  def start_link(_args) do
    GenServer.start_link(__MODULE__, [], name: __MODULE__)
  end

  def init(state) do
    {:ok, state}
  end
end
```

Now `$ mix dialyzer` will `done (passed successfully)`

### Using Elixir 1.5.0

If we change nothing but use Elixir 1.5 you do not get the warnings, following is output from running dialyzer with and without explicitly defining `child_spec`
```
... 
done (passed successfully)
```

**Comment by jeremyjh** [comment]
I believe this is issue in Elixir 1.6rc, not in `dialyzer`. Can you check if the fix to elixir-lang/elixir#7151 is causing, or resolves this?

**Comment by micahrye** [comment]
Thinks for the pointer. I installed Elixir HEAD (IEx 1.7.0-dev) and the issue is resolved.
